include/real/exact_number.hpp

```diff
@@ -693,190 +694,322 @@ namespace boost {
                 remainder = exact_remainder.digits;
             }
 
-            /// calculates *this / divisor 
-            /// 
-            ///  @brief a binary-search type method for dividing exact_numbers.
-            ///  @param is_upper true: returns result with an error of +epsilon, while
-            ///                  false: returns result with an error of -epsilon
-            void divide_vector(exact_number<T> divisor, unsigned int maximum_precision) {
-                /// @TODO: replace this with something more efficient, like newton-raphson method
-                // it also completely recalculates on each precision increase
-                // instead, could use previous information to make better "guesses"
-                // for our iteration scheme.
-
-                boost::real::exact_number<T> numerator;
-                boost::real::exact_number<T> left;
-                boost::real::exact_number<T> right;
-                boost::real::exact_number<T> residual;
-                boost::real::exact_number<T> tmp;
-                boost::real::exact_number<T> half;
-                boost::real::exact_number<T> distance;
-                boost::real::exact_number<T> min_boundary_n;
-                boost::real::exact_number<T> min_boundary_p;
-
-                bool positive = ((*this).positive == divisor.positive);
-                numerator = (*this).abs();
-                divisor = divisor.abs();
-
-                // ensuring that assignment from -1 * (maximum_precision) to exponent will not
-                // overflow
+            /*              DIVISION
+             *  @brief:  divides (*this) by divisor
+             *  @param:  divisor: number which divides (*this)
+             *  @param:  maximum_precision: Absolute Error in the result should be < 1*base^(-maximum_precision)
+             *  @param:  upper: if true: error lies in [0, +epsilon]
+             *                  else: error lies in [-epsilon, 0], here epsilon = 1*base^(-maximum_precision)
+             *  @author: Kishan Shukla
+             */
+            void divide_vector(const exact_number<T> divisor, unsigned int maximum_precision, bool upper) {
+
+                newton_raphson_division(divisor, maximum_precision, upper);
+
+            }
+
+            /*          BINARY SEARCH BASED DIVISION
+             *      an approximate division method, used in the initial guess of reciprocal in Newton Raphson
+             */
+
+            void binary_search_division(
+                const exact_number<T> divisor,
+                unsigned int maximum_precision){
+
+                /* Exceptions */
                 if (maximum_precision > (unsigned int) std::abs(std::numeric_limits<exponent_t>::min())) {
```

Comment:
I think we have already talk about the maximum_precision, but you shouldn't use it to determine where to truncate the numbers.

You should have a max_error that is a parameter of the function, then, the user of the function will define what is the max allowed error. This idea of the max_error, is the same idea we discussed in the newton division algorithm in slack that you can also implement here. Then, you can forward the max_error passed to the newton division algorithm to the binary_search_division to determine the initial guess.

Reply:
Actually, this maximum_precision is nothing but max_error you are talking about, I need to remove this exception as it doesn't make sense. Also, I'll change the name "maximum_precision" to something more correct and meaningful.